packages/gatsby/src/schema/schema.js

```diff
@@ -161,26 +169,71 @@ const addTypes = ({ schemaComposer, types, parentSpan }) => {
       })
 
       if (type) {
+        const typeName = type.getTypeName()
+        const createdFrom = `typeBuilder`
+        checkIsAllowedTypeName(typeName)
+        if (schemaComposer.has(typeName)) {
+          const typeComposer = schemaComposer.get(typeName)
+          mergeTypes({
+            typeComposer,
+            type,
+            plugin,
+            createdFrom,
+            parentSpan,
+          })
+        } else {
+          processAddedType({
+            schemaComposer,
+            type,
+            parentSpan,
+            createdFrom,
+            plugin,
+          })
+        }
+      }
+    } else {
+      const typeName = typeOrTypeDef.name
+      const createdFrom = `graphql-js`
+      checkIsAllowedTypeName(typeName)
+      if (schemaComposer.has(typeName)) {
+        const typeComposer = schemaComposer.get(typeName)
+        mergeTypes({
+          typeComposer,
+          type: typeOrTypeDef,
+          plugin,
+          createdFrom,
+          parentSpan,
+        })
+      } else {
         processAddedType({
           schemaComposer,
-          type,
+          type: typeOrTypeDef,
           parentSpan,
-          createdFrom: `typeBuilder`,
+          createdFrom,
           plugin,
         })
       }
-    } else {
-      processAddedType({
-        schemaComposer,
-        type: typeOrTypeDef,
-        parentSpan,
-        createdFrom: `graphql-js`,
-        plugin,
-      })
     }
   })
 }
 
+const mergeTypes = ({
+  typeComposer,
+  type,
+  plugin,
+  createdFrom,
+  parentSpan,
+}) => {
+  typeComposer.merge(type)
+  if (
```

Comment:
Why is it only merged for those types?

Reply:
Honestly can't remember, changed to NamedTypeComposer